src/library_dylink.js

```diff
@@ -402,9 +407,13 @@ var LibraryDylink = {
       // dlopen'ing the SIDE_MODULE.  Since we don't know the size of the explicitly initialized data
       // here, we just zero the whole thing, which is suboptimal, but should at least resolve bugs
       // from uninitialized memory.
-      for (var i = memoryBase; i < memoryBase + memorySize; i++) {
-        HEAP8[i] = 0;
-      }
+#if USE_PTHREADS
+      // in a pthread the module heap was already allocated and initialized in the main thread.
+      if (!ENVIRONMENT_IS_PTHREAD)
+#endif
+        for (var i = memoryBase; i < memoryBase + memorySize; i++) {
+          HEAP8[i] = 0;
+        }
```

Comment:
Is this assuming that the main thread is also the first thread to load the module?    I'm not sure we have decided this can/is always the case.    we might instead need some shared data structure (in shared memory) that lists all the module along with their memory base, that way when we come to load a give module we know if we are first and we can allocate and initialize the memory, otherwise we can skip it.

Reply:
The main thread doesn't need  to be the first thread to load, but it needs to load before running code using globals in any thread. AFAIK that holds true, is that right?

The main thread zero-initialize and calls the constructors. The other threads just point their `__heap_base` to the same location as the main thread, and that's why the main thread needs to load before using any globals.

The change does assume that dlopen will not load a new library (all side libraries are pre-loaded). But I think that's a general assumption with pthread + dylink at the moment.

In the long run I agree that having a structure in shared memory is a better solution (it would be more robust and scale better to cases like allowing arbitrary dlopen). I can give it a go if you think that should be included as part of this change, otherwise I am also happy to give it a go as a follow up change.

Reply:
In general I'm OK with with that happening as a followup, but I'm not a fan the change to `getMemory` as it stands.   Perhaps if we renamed `getMemory` to something more specific such as `getModuleDataAddress`  or `getOrAllocateModuleMemory` or something like that?   

As it stands `getMemory` has historically been a generic allocator and this change certainly changes that.

Reply:
Can you use opening and closing braces here and above.. to match out JS style.

Reply:
On a second thought, the shared memory approach would require some kind of locking/mutex, and that wouldn't play nicely with the main thread.

I'm somewhat confused since `__wasm_init_memory` does wait if someone else is initializing the memory. Is it guaranteed to be the main thread in this case?

An alternative would be to proxy all the operations to the main thread, so that the main thread never needs to wait, but then there's not much to gain from storing the data in shared memory.

Reply:
Imagine you load a module at runtime, you need to call malloc to allocate its memory.  Now the same module needs to be loaded on every other thread, and the base address (__memory_base) needs to be the same as it was on the main thread.  So at the very least we need to store the base address of each module in shared memory (or somehow share it between threads).  

__wasm_init_memory works I believe because all it cares about is that it runs exacltly.  Its results don't need to be shared between theads in the way the getMemory results do (IIUC).

Reply:
My confusion about `__wasm_init_memory` is about the waiting, since a shared memory solution here could use a similar strategy to ensure only one thread is writing to it. Where does the guarantee that the wait doesn't happen in the main thread come from?

The case I am thinking about is the main thread trying to load a library while another thread is already allocating the memory for it. The main thread would have to wait until the other thread finishes allocating (and initializing) that memory so that it can use the same address. This wait operation might not always be possible (e.g., browser render thread). I believe this is the same challenge faced by `__wasm_init_memory`, but I don't understand why it works in that case.

Is this PR the right place for this discussion? Or should I move this to a different forum?

Reply:
I think the reason `__wasm_init_memory` never blocks on the main thread today is simply that the main thread always instantiates the module before postMessaging it to background threads, so it always wins the race.

I think that the synchronization of the various module across all the threads is the single biggest challenge with pthreads + dynamic linking.   There are many unanswered questions and a lot of them relate to how we do the loading (and potentially the blocking) on different threads.   Perhaps the best place to discuss these kinds of issues is on #3494?